Sources/SwiftSyntax/RawSyntax.swift

```diff
@@ -1274,65 +553,252 @@ extension RawSyntax {
+  let totalTriviaCount = leadingTrivia.count + trailingTrivia.count
+
+  if totalTriviaCount != 0 {
+    var byteLength = 0
+    let buffer = arena.allocateRawTriviaPieceBuffer(count: totalTriviaCount)
+    var ptr = buffer.baseAddress!
+    for piece in leadingTrivia + trailingTrivia {
+      byteLength += piece.sourceLength.utf8Length
+      ptr.initialize(to: .make(arena: arena, piece))
+      ptr = ptr.advanced(by: 1)
+    }
+    return (pieces: .init(buffer), byteLength: byteLength)
+  } else {
+
+    return (pieces: .init(start: nil, count: 0), byteLength: 0)
+  }
+}
+
+// MARK: - Factories.
+
+extension RawSyntax {
+  /// "Designated" factory method to create a materialized token node.
+  ///
+  /// This should not be called directly.
+  /// Use `makeMaterializedToken(arena:kind:leadingTrivia:trailingTrivia:)` or
+  /// `makeEmptyToken(arena:kind:)` instead.
+  ///
+  /// - Parameters:
+  ///   - arena: SyntaxArea to the result node data resides.
+  ///   - kind: Token kind.
+  ///   - text: Token text.
+  ///   - triviaPieces: Raw trivia pieces including leading and trailing trivia.
+  ///   - numLeadingTrivia: Number of leading trivia pieces in `triviaPieces`.
+  ///   - byteLength: Byte length of this token including trivia.
+  internal static func materializedToken(
+    arena: SyntaxArena,
+    kind: RawTokenKind,
+    text: SyntaxText,
+    triviaPieces: RawTriviaPieceBuffer,
+    numLeadingTrivia: UInt32,
+    byteLength: UInt32
+  ) -> RawSyntax {
+    let payload = RawSyntaxData.MaterializedToken(
+      tokenKind: kind, tokenText: text,
+      triviaPieces: triviaPieces,
+      numLeadingTrivia: numLeadingTrivia,
+      byteLength: byteLength)
+    return RawSyntax(arena: arena, payload: .materializedToken(payload))
+  }
+
+  /// Factory method to create a materialized token node.
+  ///
+  /// - Parameters:
+  ///   - arena: SyntaxArea to the result node data resides.
+  ///   - kind: Token kind.
+  ///   - text: Token text.
+  ///   - leadingTrivia: Leading trivia.
+  ///   - trailingTrivia: Trailing trivia.
+  static func makeMaterializedToken(
+    arena: SyntaxArena,
+    kind: TokenKind,
+    leadingTrivia: Trivia,
+    trailingTrivia: Trivia
+  ) -> RawSyntax {
+    let decomposed = kind.decomposeToRaw()
+    let rawKind = decomposed.rawKind
+    let text: SyntaxText = (decomposed.string.map({arena.intern($0)}) ??
+                            decomposed.rawKind.defaultText ??
+                            "")
+
+    var byteLength = text.count
+
+    let triviaPieces = makeRawTriviaPieces(
+      arena: arena, leadingTrivia: leadingTrivia, trailingTrivia: trailingTrivia)
+
+    byteLength += triviaPieces.byteLength
+
+    return .materializedToken(
+      arena: arena, kind: rawKind, text: text,
+      triviaPieces: triviaPieces.pieces,
+      numLeadingTrivia: numericCast(leadingTrivia.count),
+      byteLength: numericCast(byteLength))
+  }
+
+  static func makeEmptyToken(
```

Comment:
Do we want to call this `makeEmptyToken` instead of `makeMissingToken`?

Reply:
This naming is an artifact of my local development. I'll rename it